zim/gui/pageview/__init__.py

```diff
@@ -2342,8 +2344,18 @@ def foreach_line_in_selection(self, func, *args, **kwarg):
 				# exclude last line if selection ends at newline
 				# because line is not visually part of selection
 				end.backward_char()
-			for line in range(start.get_line(), end.get_line() + 1):
-				func(line, *args, **kwarg)
+			if skip_empty_lines:
+				all_lines_are_empty = True
+				for line in range(start.get_line(), end.get_line() + 1):
+					if not self.get_line_is_empty(line):
+						all_lines_are_empty = False
+						func(line, *args, **kwarg)
+				if all_lines_are_empty:
```

Comment:
I feel like a second kwarg that decides whether to call the func on empty lines iff all_lines_are_empty would be a bit excessive. On the other hand, it's the cleanest solution to support both the behavior you described (the user called format on empty lines, so we format the empty lines) as well as the previous behavior of `toggle_bullet`